### ISSUE TITLE
System Design: Ratelimiter

### ISSUE THREAD

**Issue by hsunsmile** [hsunsmile/hsunsmile.github.io#15]
9. Design an API Rate Limiter (e.g. for Firebase or Github)

You are expected to develop a Rate Limiter services that can:

    Limit the number of requests an entity can send to an API within a time window e.g., 15 requests per second.
    The rate limiting should work for a distributed setup, as the APIs are accessible through a cluster of servers.
    How would you handle throttling (soft and hard throttling etc.).

**Comment by hsunsmile** [comment]
## TLDR
- understand what is the most interesting part to the interviewer (enforcement? or other parts)
- learn token bucket algorithm
- Is losing current count important? (fail open)

#### What is ratelimiter
Noisy neighbors
Autoscaling can take time

#### Stores rules in MySQL
If it is 5 min then we do not have to discuss about MySQL LB

Ratelimit on something (IP, userAgent, user_id, etc)

#### How to enforce the policy
Token bucket algorithms
Fail open if the service dead (sidecar if K8S)

Rules format
- name
- IP
- time_range: s/min/hour
- count: API request within the time range meeting the condition

How many rules
Analytics good to have
- logs to datalake
- metric servers (Datadag)
- metric names

#### System diagram
<img width="632" alt="image" src="https://user-images.githubusercontent.com/736241/93555603-4a14ab80-f92b-11ea-8653-578c2c6637d6.png">

**Comment by hsunsmile** [comment]
probability based distributed rate-limiter. https://www.ably.io/blog/distributed-rate-limiting-scale-your-platform/
